Hashcash.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Hashcash 3"
.TH Hashcash 3 "2005-08-02" "perl v5.8.5" "User Contributed Perl Documentation"
.SH "NAME"
Digest::Hashcash \- generate Hashcash stamps (http://www.hashcash.org)
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& use Digest::Hashcash;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This module implements the hashcash hash (or digest, although it's
not clearly a digest). For all your information needs please visit
http://www.hashcash.org.
.PP
One thing to note about this module is that it requires \s-1ISO\s0 C99
support, both in your compiler and your standard library.  If you
don't have a compiler that supports \s-1ISO\s0 C, get gcc at
http://gcc.gnu.org/ :)
.ie n .IP "$secs = estimate_time $size" 4
.el .IP "$secs = estimate_time \f(CW$size\fR" 4
.IX Item "$secs = estimate_time $size"
Estimate the average time necessary to calculate a token of the given
size.
.Sp
See also \f(CW\*(C`estimate_size\*(C'\fR.
.ie n .IP "$size = estimate_size $time\fR[, \f(CW$min]" 4
.el .IP "$size = estimate_size \f(CW$time\fR[, \f(CW$min\fR]" 4
.IX Item "$size = estimate_size $time[, $min]"
Estimate the size that can be calculated in the given time (which is an
upper bound). The function will not return a size less then \f(CW\*(C`min\*(C'\fR.
.Sp
Estimating the time to be used can go wrong by as much as 50% (but is
usually quite accurate), and the estimation itself can take as much as a
second on slower (<pentium) machines, but faster machines (1Ghz P3 for
example) usually handle it within a hundredth of a second or so.
.Sp
The estimation will be done only once, so you can call this fucntion as
often as you like without incuring the overhead everytime.
.IP "$cipher = new Digest::Hashcash [param => value...]" 4
.IX Item "$cipher = new Digest::Hashcash [param => value...]"
.RS 4
.PD 0
.IP "size => 20" 4
.IX Item "size => 20"
.PD
The number of collisions, in bits. Every bit increases the time to create
the token (and thus the cash) by two.
.IP "vers => 1" 4
.IX Item "vers => 1"
Default version 1.  Can produce version 0 if required for backwards
compatibility.
.ie n .IP "uid => """"" 4
.el .IP "uid => ``''" 4
.IX Item "uid => """""
A string used to make the token more unique (e.g. the senders address)
and reduce token collisions. The string must only contain characters
valid for the trial part of the token, e.g. uuencoded, base64 or
e\-mail-address-parts are useful here.  Deprecated: use extension field
if required.
.IP "extrarand => 0" 4
.IX Item "extrarand => 0"
The extra bytes of randomness to add to the token in addition to the
standard amount. Each byte adds a little bit over 6 bit of randomness to
the token.
.Sp
The standard amount of randomness is 8 (> 51 bits of randomness).
.IP "timestamp => 0" 4
.IX Item "timestamp => 0"
The timestamp to use. A value of 0 (the default) means to use the current
time.
.RE
.RS 4
.RE
.ie n .IP "$token = $cipher\->hash ($data [, param => value...])" 4
.el .IP "$token = \f(CW$cipher\fR\->hash ($data [, param => value...])" 4
.IX Item "$token = $cipher->hash ($data [, param => value...])"
Creates and returns a new token. This can take some time.
.Sp
Any additional parameters are interpreted the same way as arguments to
\&\f(CW\*(C`new\*(C'\fR.
.ie n .IP "$prefix = $cipher\->verify ($token)" 4
.el .IP "$prefix = \f(CW$cipher\fR\->verify ($token)" 4
.IX Item "$prefix = $cipher->verify ($token)"
Version 0: Checks the given token and returns the number of collision
bits.
.Sp
Version 1: Returns 0 if stated value is more than the computed
collision value, otherwise returns the stated stamp value.
.Sp
Any additional parameters are interpreted the same way as arguments to
\&\f(CW\*(C`new\*(C'\fR.
.ie n .IP "$version = $cipher\->version ($token)" 4
.el .IP "$version = \f(CW$cipher\fR\->version ($token)" 4
.IX Item "$version = $cipher->version ($token)"
Returns the version of the stamp (currently 0 or 1).
.ie n .IP "$resource = $cipher\->resource ($token)" 4
.el .IP "$resource = \f(CW$cipher\fR\->resource ($token)" 4
.IX Item "$resource = $cipher->resource ($token)"
Returns the resource part, or \f(CW\*(C`undef\*(C'\fR.
.ie n .IP "$tstamp = $cipher\->timestamp ($token)" 4
.el .IP "$tstamp = \f(CW$cipher\fR\->timestamp ($token)" 4
.IX Item "$tstamp = $cipher->timestamp ($token)"
Returns the timestamp part (in the same format as perls \f(CW\*(C`time\*(C'\fR), or
\&\f(CW\*(C`undef\*(C'\fR.
.ie n .IP "$extension = $cipher\->extension ($token [,$name [,$var]])" 4
.el .IP "$extension = \f(CW$cipher\fR\->extension ($token [,$name [,$var]])" 4
.IX Item "$extension = $cipher->extension ($token [,$name [,$var]])"
For Version 1 stamps returns the extension part; for Version 0 stamps
returns undef.  In a scalar context returns the extension string, in
array context returns associative array containing extensions as keys
and their values as the corresponding value.  If the optional name
argument is given, returns info but about that extension (ie decodes
the options of that extension if called in array context); finally if
the optional var argument also is given reports just the (scalar)
value of that variable in the named extension.
.Sp
Note it is valid for an extension to exist (the key exists in the
associative array) but to have undefined value, this corresponds to a
boolean option without a value.  Like \*(L"noreply;foo=bar\*(R" has two
extensions, noreply and foo; noreply has no associated value, so the
method extension would parse such a stamp into an associative array
containing key \*(L"noreply\*(R" with an undefined value; and key \*(L"foo\*(R"
containing associated value \*(L"bar\*(R".
.SH "BUGS"
.IX Header "BUGS"
.Vb 5
\& * There is a y2k+100 problem, as I always assume the same as
\&   Time::Local.  This is a problem with the hashcash specification,
\&   which specifies years as 2 digits :( Though it hardly matters --
\&   after 100 years of Moore's law a 20 bit stamp will be tiny and not
\&   worth storing.
.Ve
.PP
.Vb 2
\& * extension method could be more efficient (it does not cache its
\&   parsed results so if used in a loop reparses on each call)
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
.Vb 2
\& Marc Lehmann <schmorp@schmorp.de>
\& http://home.schmorp.de
.Ve
.PP
.Vb 2
\& Adam Back <adam@cypherspace.org> added version 1 support
\& http://www.cypherspace.org/adam/
.Ve


Changes
Revision history for Perl extension Digest::Hashcash

0.04  Tue Aug  2 2005
	- changes from Jason to verify version 1 stamps
	- more changes from Adam Back: code to create version 1 stamps;
	  associated methods for v1 stamps: extension(), version()
	- create v1 stamps by default
	- fix some bugs (hash method was ignoring arguments)
	- add extension parsing
	- fix randomness (earlier versions using rand(3))
	- add version1 extension tests

0.03  Thu Mar  3 18:00:52 CET 2005
	- generate more efficient code with gcc-3.4 and later.
        - change of contact address.

0.02  Thu Sep 11 08:16:54 CEST 2003
        - some workarounds for antiquated compilers like gcc-2.95.

0.01  Sat Sep  6 17:10:40 CEST 2003
	- original version; cloned from Crypt::Twofish2


Hashcash.pm
=head1 NAME

Digest::Hashcash - generate Hashcash stamps (http://www.hashcash.org)

=head1 SYNOPSIS

 use Digest::Hashcash;

=head1 DESCRIPTION

This module implements the hashcash hash (or digest, although it's
not clearly a digest). For all your information needs please visit
http://www.hashcash.org.

One thing to note about this module is that it requires ISO C99
support, both in your compiler and your standard library.  If you
don't have a compiler that supports ISO C, get gcc at
http://gcc.gnu.org/ :)

=over 4

=cut

package Digest::Hashcash;

use Time::Local;
use Time::HiRes;

require XSLoader;

no warnings;

$VERSION = 0.04;

XSLoader::load Digest::Hashcash, $VERSION;

=item $secs = estimate_time $size

Estimate the average time necessary to calculate a token of the given
size.

See also C<estimate_size>.

=item $size = estimate_size $time[, $min]

Estimate the size that can be calculated in the given time (which is an
upper bound). The function will not return a size less then C<min>.

Estimating the time to be used can go wrong by as much as 50% (but is
usually quite accurate), and the estimation itself can take as much as a
second on slower (<pentium) machines, but faster machines (1Ghz P3 for
example) usually handle it within a hundredth of a second or so.

The estimation will be done only once, so you can call this fucntion as
often as you like without incuring the overhead everytime.

=cut

my $rounds;

sub _rounds {
   $rounds ||= &_estimate_rounds();
}

sub estimate_time {
   my ($size) = @_;
   2**$size / &_rounds;
}

sub estimate_size {
   my ($time, $min) = @_;
   $time = (log $time * $rounds) / log 2;
   $time < $min ? $min : int $time;
}

=item $cipher = new Digest::Hashcash [param => value...]

=over 4

=item size => 20

The number of collisions, in bits. Every bit increases the time to create
the token (and thus the cash) by two.

=item vers => 1

Default version 1.  Can produce version 0 if required for backwards
compatibility.

=item uid => ""

A string used to make the token more unique (e.g. the senders address)
and reduce token collisions. The string must only contain characters
valid for the trial part of the token, e.g. uuencoded, base64 or
e-mail-address-parts are useful here.  Deprecated: use extension field
if required.

=item extrarand => 0

The extra bytes of randomness to add to the token in addition to the
standard amount. Each byte adds a little bit over 6 bit of randomness to
the token.

The standard amount of randomness is 8 (> 51 bits of randomness).

=item timestamp => 0

The timestamp to use. A value of 0 (the default) means to use the current
time.

=back

=item $token = $cipher->hash ($data [, param => value...])

Creates and returns a new token. This can take some time.

Any additional parameters are interpreted the same way as arguments to
C<new>.

=item $prefix = $cipher->verify ($token)

Version 0: Checks the given token and returns the number of collision
bits.

Version 1: Returns 0 if stated value is more than the computed
collision value, otherwise returns the stated stamp value.

Any additional parameters are interpreted the same way as arguments to
C<new>.

=item $version = $cipher->version ($token)

Returns the version of the stamp (currently 0 or 1).

=item $resource = $cipher->resource ($token)

Returns the resource part, or C<undef>.

=item $tstamp = $cipher->timestamp ($token)

Returns the timestamp part (in the same format as perls C<time>), or
C<undef>.

=item $extension = $cipher->extension ($token [,$name [,$var]])

For Version 1 stamps returns the extension part; for Version 0 stamps
returns undef.  In a scalar context returns the extension string, in
array context returns associative array containing extensions as keys
and their values as the corresponding value.  If the optional name
argument is given, returns info but about that extension (ie decodes
the options of that extension if called in array context); finally if
the optional var argument also is given reports just the (scalar)
value of that variable in the named extension.

Note it is valid for an extension to exist (the key exists in the
associative array) but to have undefined value, this corresponds to a
boolean option without a value.  Like "noreply;foo=bar" has two
extensions, noreply and foo; noreply has no associated value, so the
method extension would parse such a stamp into an associative array
containing key "noreply" with an undefined value; and key "foo"
containing associated value "bar".

=back

=cut

sub encode {
    my %ext = shift;
    my $res;
    foreach $key (keys %ext) {
	if ( index( $ext[$key], "," ) >= 0 ) {
	    $ext[$key] = split( ",", $ext[$key] );
	}
    }
}

sub new {
   my $class = shift;

   bless { @_ }, $class;
}

sub hash {
   my $self = shift;
   my $resource = shift;
   my %arg = ( vers => 1, %$self, resource => $resource, @_ );

   &_gentoken(@arg{qw(size vers timestamp resource extension uid extrarand)});
}

sub verify {
   my ($self, $token) = (shift, shift);

   my $prefix = &_prefixlen($token);

   if ($token =~ /^0:/) {
      return $prefix >= 0 ? $prefix : 0;
   } elsif ($token =~ /^1:/) {
      ($ver, $bits, $ts, $res, $ext, $junk, $count) = split(':', $token, 7);
      return $prefix >= $bits ? $bits : 0;
   }
   else { return undef; }
}

sub resource {
   my ($self, $token) = @_;

   if ($token =~ /^0:/) {
      ($ver, $ts, $res, $rand) = split(':', $token, 4);
   }
   elsif ($token =~ /^1:/) {
      ($ver, $bits, $ts, $res, $ext, $junk, $count) = split(':', $token, 7);
   }
   else { return undef; }
   return $res;
}

sub timestamp {
   my ($self, $token) = @_;

   if ($token =~ /^0:/) {
      ($ver, $ts, $res, $rand) = split(':', $token, 4);
   }
   elsif ($token =~ /^1:/) {
      ($ver, $bits, $ts, $res, $ext, $junk, $count) = split(':', $token, 7);
   }
   else { return undef; }

   my ($y, $m, $d, $H, $M, $S);
   local $_ = $ts;
   $y = /\G(\d\d)/gc ? $1 : return undef;
   $m = /\G(\d\d)/gc ? $1 : 1;
   $d = /\G(\d\d)/gc ? $1 : 1;
   $H = /\G(\d\d)/gc ? $1 : 0;
   $M = /\G(\d\d)/gc ? $1 : 0;
   $S = /\G(\d\d)/gc ? $1 : 0;

   return timegm $S, $M, $H, $d, $m - 1, $y;
}

sub extension { 
   my ($self, $token, $name, $var) = @_;
   my $bits, $ts, $res, $ext, $junk, $count, @ext, %ext;

   if ($token =~ /^0:/) {
      return undef;
   }
   elsif ($token =~ /^1:/) {
       if ( wantarray ) { %ext = (); }
      ($ver, $bits, $ts, $res, $ext, $junk, $count) = split(':', $token, 7);
      if ( wantarray || defined( $name ) ) {
	  foreach $pair ( split( ";", $ext ) ) {
	      ($nam,$val) = split( "=", $pair, 2 );
	      if ( defined( $name ) ) {
		  if ( $name eq $nam ) {
		      if ( wantarray || defined( $var ) ) {
			  foreach $vpair ( split( ",", $val ) ) {
			      ($vnam,$vval) = split( "=", $vpair, 2 );
			      if ( defined( $var ) ) {
				  if ( $vnam eq $var ) { return $vval; }
			      } else {
				  $ext{$vnam} = $vval;
			      }
			  }
			  return %ext;
		      } else {
			  return $val;
		      }
		  }
	      } else {
		  $ext{$nam} = $val;
	      }
	  }
	  return %ext;
      } else {
	  return $ext;
      }
   }
   else { return undef; }
}

sub version {
   my ($self, $token) = @_;

   if ($token =~ /^0:/) {
      return 0;
   } elsif ($token =~ /^1:/) {
      return 1;
   }
   else { return undef; }
}
=head1 SEE ALSO

L<http://www.hashcash.org>.

=head1 BUGS

 * There is a y2k+100 problem, as I always assume the same as
   Time::Local.  This is a problem with the hashcash specification,
   which specifies years as 2 digits :( Though it hardly matters --
   after 100 years of Moore's law a 20 bit stamp will be tiny and not
   worth storing.

 * extension method could be more efficient (it does not cache its
   parsed results so if used in a loop reparses on each call)

=head1 AUTHOR

 Marc Lehmann <schmorp@schmorp.de>
 http://home.schmorp.de

 Adam Back <adam@cypherspace.org> added version 1 support
 http://www.cypherspace.org/adam/

=cut

1;
